### ISSUE TITLE
ExEx WAL inconsistency

### ISSUE THREAD

**Issue by shekhirin** [paradigmxyz/reth#11168]
## Problem

In the current design, we will delete the files containing notifications from the ExEx's WAL directory immediately on rollback. This may be unsafe in the following scenario:
1. We detect that the ExEx is on a wrong fork. It means that its chain needs to be first reverted to the first canonical block, and then a new chain from the canonical fork should be committed with.
2. We roll back the ExEx's WAL, removing the files with notifications from the disk.
3. We generate reverts from the notifications that were rolled back send them to the ExEx.
4. Node shuts down before the ExEx finished processing the reverts of rolled back notifications.
5. Node is back up, the ExEx is still not on the canonical fork, but we also have no non-canonical notifications in the WAL because we deleted them

## Solution

We need to delete notifications from the WAL only after ExEx actually processed them. There may be a couple of options here:
1. Do not delete notifications from the WAL on rollback and instead write reverts of these notifications to the WAL. If the node crashes, we still have the chain that needs to be rolled back. The rolled-back chain will be deleted on finalization.
2. We should be aware of the `FinishedHeight` for a particular ExEx and only delete those rolled back notifications from the WAL that ExEx processed the reverts for.

IMO the first approach is easier, let's look at an example:
1. ExEx processes blocks `A, B, C`, but the blocks `B` and `C` aren't committed to the node database yet.
2. ExEx shuts down and the node loses blocks `B` and `C` because they were in memory.
3. ExEx is back, and during the downtime node's blocks `B` and `C` were reorged into `B'` and `C'` respectively. The ExEx is still on the chain `A, B, C`.
4. ExEx creates a new subscription with the head `C`. We detect that it's not canonical and walk the WAL to the first canonical block, which is `A`. We create reverts of `C` and `B` and write them to the ExEx's WAL. Then we send these reverts to the ExEx.
5. ExEx processes the revert of `C`, but exits before processing the revert of `B`.
6. ExEx starts again and creates a new subscription with the head `B` (because the `C` was already reverted). We detect that it's not canonical and walk the WAL to the first canonical block, which is `A`. We need to create a revert of `B`, but before that we check if there's already one in the WAL. There is, so we just grab it and send it to the ExEx.
7. ExEx processes the revert of `B`, and we can start sending canonical notifications for `B'` and `C'`.

**Comment by shekhirin** [comment]
Superseded by https://github.com/paradigmxyz/reth/issues/11202
